text/0000-process-stdio-redirection.md

```diff
@@ -0,0 +1,106 @@
+- Feature Name: process-stdio-redirection
+- Start Date: 2015-04-10
+- RFC PR:
+- Rust Issue:
+
+# Summary
+
+Update the standard library with a cross-platform high level API for redirecting
+stdio of child processes to any opened file handle or equivalent.
+
+# Motivation
+
+The current API in `std::process` allows the usage of raw file descriptors or
+HANDLEs for child process stdio redirection by leveraging the `FromRaw{Fd,
+Handle}` and `AsRaw{Fd, Handle}` trait implementations on `Stdio` and
+`ChildStd{In, Out, Err}`, respectively.
+
+Unfortunately, since the actual methods pertaining to `FromRaw*` and `AsRaw*`
+are OS specific, their usage requires either constant `cfg` checks or OS
+specific code duplication. Moreover, the conversion to `Stdio` is `unsafe` and
+requires the caller to ensure the OS handle remains valid until spawning the
```

Comment:
@alexcrichton 
>the contract of FromRawFd is that it takes ownership of the file descriptor (or handle) in question

Ah, I had overlooked this...

However, doesn't this require that the caller `mem::forget` the wrapper over the descriptor/HANDLE or at least manually `dup` it (and perform error checking)? Because the contract of `AsRawFd` does *not* transfer ownership and then both `Stdio` and the original wrapper will both try to close the handle.

>I suspect that adding traits like `IntoRawFd` are somewhat inevitable.

I also agree, however, I feel like something like `IntoRawFd` could be very useful now, because being forced to `mem::forget` or `dup` handles (i.e. when using the current state of the API) is *very* unergonomic to force on userland, especially when the standard library does such a good job from abstracting much of that away (and the reason why the `Command` interface is so handy to have).

>The alternative (`IntoRawFd`) I don't think is unergonomic *enough* to warrant a specialized function.

I could definitely live with a compromise like `IntoRawFd`, but a specialized function could avoid accidental use of `AsRawFd` instead of `IntoRawFd` and prevent someone from shooting their foot (since both traits would produce a fd which `FromRawFd` will happily accept).

Assuming the availability of an `IntoRawFd` trait the currently proposed API could be tweaked to use it more appropriately:
```rust
pub trait StdioExt {
   // Take ownership of the handle
   fn redirect<T: IntoRaw*>(t: T) -> Stdio;
   // Unsafely borrow the handle, letting caller ensure it is valid
   unsafe fn redirect_by_ref<T: AsRaw*>(t: &T) -> Stdio;
}
```